src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -211,7 +211,7 @@ public override BoundStatement InstrumentReturnStatement(BoundReturnStatement or
 
             // A synthesized return statement that does not return a value never requires instrumentation.
             // A property set method defined without a block has such a synthesized return statement.
-            if (!_methodHasExplicitBlock && ((BoundReturnStatement)original).ExpressionOpt != null)
+            if ((!_methodHasExplicitBlock || _methodBodyFactory.CurrentMethod.MethodKind == MethodKind.LambdaMethod) && ((BoundReturnStatement)original).ExpressionOpt != null)
```

Comment:
>_methodBodyFactory.CurrentMethod.MethodKind == MethodKind.LambdaMethod [](start = 45, length = 70)

This feels fragile to me. We are relying on the state of the _methodBodyFactory, which isn't maintained by this component itself. I think the check could be purely syntactical, we could try to check if the syntax associated with the expression is a body of a lambda instead.